src/test/java/org/jenkinsci/plugins/durabletask/BourneShellScriptTest.java

```diff
@@ -432,50 +458,95 @@ private Slave prepareDockerPlatforms() throws Exception {
             }
             Thread.sleep(100);
         }
-        System.out.println("output: " + baos);
         assertEquals(0, c.exitStatus(ws, launcher, listener).intValue());
         assertTrue(baos.toString().contains("hello world"));
         c.cleanup(ws);
-        assertThat(getZombies(), isEmptyString());
+        assertNoZombies();
+    }
+
+    @Test public void binaryCaching() throws Exception {
+        assumeTrue(!platform.equals(TestPlatform.UBUNTU_NO_BINARY));
+        String os;
+        switch (platform) {
+            case NATIVE:
+                if (Platform.isDarwin()) {
+                    os = "darwin";
+                } else {
+                    os = "unix";
+                }
+                break;
+            default:
+                os = "unix";
+        }
+
+        String version = j.getPluginManager().getPlugin("durable-task").getVersion();
+        version = StringUtils.substringBefore(version, "-");
+        String binaryName = "durable_task_monitor_" + version + "_" + os + "_64";
+        FilePath binaryPath = ws.getParent().getParent().child("caches/durable-task/" + binaryName);
+        assertFalse(binaryPath.exists());
+
+        BourneShellScript script = new BourneShellScript("echo hello");
+        EnvVars envVars = new EnvVars();
+        Controller c = script.launch(envVars, ws, launcher, listener);
+        awaitCompletion(c);
+        assertEquals(0, c.exitStatus(ws, launcher, listener).intValue());
+        assertTrue(binaryPath.exists());
+        Long timeCheck1 = binaryPath.lastModified();
+
+        c = script.launch(envVars, ws, launcher, listener);
+        awaitCompletion(c);
+        assertEquals(0, c.exitStatus(ws, launcher, listener).intValue());
+        Long timeCheck2 = binaryPath.lastModified();
+        assertEquals(timeCheck1, timeCheck2);
+
+        binaryPath.delete();
+        binaryPath.touch(Instant.now().toEpochMilli());
+        try {
+            c = script.launch(envVars, ws, launcher, listener);
+        } catch (Exception e) {
+            assertThat(e, instanceOf(IOException.class));
```

Comment:
To make sure that the try always throws an exception and hits this case, you should call `Assert.fail` at the end of the try. Right now, this test will pass if `script.launch` succeeds (it shouldn't since you corrupted the cache) _or_ if it throws the specific kind of exception you are looking for.

Reply:
Ah you're right. If I had done caching incorrectly and the binary copied over, it would've passed.